pubsignals/common.go

```diff
@@ -0,0 +1,248 @@
+	for k, v := range obj {
+		entries[k] = v
+	}
+	return entries
+}
+
+// ParseQueryMetadata parse property query and return query metadata
+func ParseQueryMetadata(ctx context.Context, propertyQuery PropertyQuery, ldContextJSON, credentialType string, options merklize.Options) (query *QueryMetadata, err error) {
+	datatype, err := options.TypeFromContext([]byte(ldContextJSON), fmt.Sprintf("%s.%s", credentialType, propertyQuery.FieldName))
+	if err != nil {
+		return nil, err
+	}
+
+	query = &QueryMetadata{
+		PropertyQuery:   propertyQuery,
+		SlotIndex:       0,
+		MerklizedSchema: false,
+		Datatype:        datatype,
+		ClaimPathKey:    big.NewInt(0),
+		Values:          []*big.Int{},
+		Path:            &merklize.Path{},
+	}
+
+	var ctxObj map[string]interface{}
+	err = json.Unmarshal([]byte(ldContextJSON), &ctxObj)
+	if err != nil {
+		return nil, err
+	}
+
+	ldCtx, err := ld.NewContext(nil, nil).Parse(ctxObj["@context"])
+	if err != nil {
+		return nil, err
+	}
+
+	serAttr, err := getSerializationAttrFromParsedContext(ldCtx, credentialType)
+	if err != nil {
+		return nil, err
+	}
+	if serAttr == "" {
+		query.MerklizedSchema = true
+	}
+
+	if !query.MerklizedSchema {
+		query.SlotIndex, err = parser.Parser{}.GetFieldSlotIndex(propertyQuery.FieldName, credentialType, []byte(ldContextJSON))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		path := merklize.Path{}
+		if query.FieldName != "" {
+			path, err = options.FieldPathFromContext([]byte(ldContextJSON), credentialType, propertyQuery.FieldName)
+			if err != nil {
+				return nil, err
+			}
+			err = path.Prepend(credentialSubjectFullKey)
+			if err != nil {
+				return nil, err
+			}
+
+			query.ClaimPathKey, err = path.MtEntry()
+			if err != nil {
+				return nil, err
+			}
+			query.Path = &path
+		}
+
+		if propertyQuery.OperatorValue != nil {
+			if !IsValidOperation(datatype, propertyQuery.Operator) {
+				return nil, fmt.Errorf("operator %d is not supported for datatype %s", propertyQuery.Operator, datatype)
+			}
+		}
+
+		query.Values, err = transformQueryValueToBigInts(ctx, propertyQuery.OperatorValue, datatype)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return query, err
+}
+
+// ParseQueriesMetadata parse credential subject and return array of query metadata
+func ParseQueriesMetadata(ctx context.Context, credentialType, ldContextJSON string, credentialSubject any, options merklize.Options) (out []QueryMetadata, err error) {
+	queriesMetadata, err := ParseCredentialSubject(ctx, credentialSubject)
+	if err != nil {
+		return nil, err
+	}
+	out = []QueryMetadata{}
+	for i := 0; i < len(queriesMetadata); i++ {
+		queryMetadata, err := ParseQueryMetadata(ctx, queriesMetadata[i], ldContextJSON, credentialType, options)
+		if err != nil {
+			return nil, err
+		}
+		out = append(out, *queryMetadata)
+	}
+	return out, err
+}
+
+func transformQueryValueToBigInts(_ context.Context, value any, ldType string) (out []*big.Int, err error) {
```

Comment:
we should fix for IN / NIN as in js

Reply:
this fix is outside `transformQueryValueToBigInts` function, maybe a good idea to create separate PR for IN / NIN ?